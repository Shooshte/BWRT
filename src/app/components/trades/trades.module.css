.list {
  background: #f3f3f3;
  border-radius: 12px;
  overflow: auto;
  padding: 1rem;
  width: calc(100% - 2rem);
}

.listItem {
  display: grid;
  grid-template-columns: max-content auto max-content;
  font-size: 12px;
  font-weight: 400;
  line-height: 21px;
  justify-content: space-between;
}

.value {
  font-weight: 600;
}
