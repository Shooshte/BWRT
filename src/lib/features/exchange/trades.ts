import type { Trade } from "./exchangeSlice";

export const mockTrades: Trade[] = [
  {
    amount: 0.00018,
    id: "4",
    price: 1063670,
    timestamp: 1738342982301,
    type: "buy",
  },
  {
    amount: 0.002,
    id: "7",
    price: 1060670,
    timestamp: 1738342981101,
    type: "sell",
  },
  {
    amount: 0.001,
    id: "9",
    price: 1063710,
    timestamp: 1738342962301,
    type: "buy",
  },
  {
    amount: 0.00017,
    id: "3",
    price: 1064000,
    timestamp: 1738326995301,
    type: "buy",
  },
  {
    amount: 0.00012,
    id: "2",
    price: 1075550,
    timestamp: 1738326984301,
    type: "sell",
  },
  {
    amount: 0.00015,
    id: "1",
    price: 1075120,
    timestamp: 1738326982301,
    type: "buy",
  },
  {
    amount: 0.0001,
    id: "5",
    price: 1075022,
    timestamp: 1738326924101,
    type: "sell",
  },
];
