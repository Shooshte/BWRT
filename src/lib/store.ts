import { configureStore } from "@reduxjs/toolkit";
import { setupListeners } from "@reduxjs/toolkit/query";
import chartReducer from "./features/chart/chartSlice";
import { coingeckoApi } from "./features/chart/coingeckoApi";

export const makeStore = () => {
  return configureStore({
    reducer: {
      chart: chartReducer,
      [coingeckoApi.reducerPath]: coingeckoApi.reducer,
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(coingeckoApi.middleware),
  });
};

setupListeners(makeStore().dispatch);

export type AppStore = ReturnType<typeof makeStore>;
export type RootState = ReturnType<AppStore["getState"]>;
export type AppDispatch = AppStore["dispatch"];
