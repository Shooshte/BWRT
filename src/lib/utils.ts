export function round(value: number, decimals: number) {
  // @ts-expect-error - adding e to a number is valid
  return Number(Math.round(value + "e" + decimals) + "e-" + decimals);
}

export function formatCurrency(value: number) {
  const [integer, decimal] = value.toString().split(".");
  const cleanedInteger = integer.replace("-", "");

  const start = cleanedInteger.length % 3;

  let integerString = "";

  if (cleanedInteger.length <= 3) {
    integerString = cleanedInteger;
  } else {
    integerString = cleanedInteger.slice(0, start);
    if (start > 0) {
      integerString += ",";
    }
    const remainingInteger = cleanedInteger.slice(start, cleanedInteger.length);

    for (let i = 0; i < remainingInteger.length; i++) {
      if (i % 3 === 0 && i > 0) {
        integerString += ",";
      }

      integerString += remainingInteger.at(i);
    }
  }

  if (cleanedInteger.length !== integer.length) {
    integerString = "-" + integerString;
  }

  if (decimal) {
    return `${integerString}.${decimal}`;
  }

  return integerString;
}
