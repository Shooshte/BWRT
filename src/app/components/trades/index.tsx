"use client";

import React, { useMemo } from "react";
import { useSelector } from "react-redux";
import { RootState } from "../../../lib/store";
import { round } from "../../../lib/utils";

import styles from "./trades.module.css";

interface ParsedTrade {
  id: string;
  time: string;
  type: string;
  value: string;
}

export default function Trades() {
  const trades = useSelector((state: RootState) => state.exchange.trades);

  const parsedTrades: ParsedTrade[] = useMemo(() => {
    return trades.map(({ amount, id, price, timestamp, type }) => {
      const roundedAmount = round(amount, 5);
      const roundedPrice = round(price * amount, 2);

      const value =
        type === "buy"
          ? `${roundedAmount} BTC / - ${roundedPrice} USD`
          : `- ${roundedAmount} BTC / + ${roundedPrice} USD`;
      const time = new Date(timestamp).toLocaleTimeString();
      return {
        id,
        time,
        type,
        value,
      };
    });
  }, [trades]);

  return (
    <ul className={styles.list}>
      {parsedTrades.map(({ id, time, type, value }) => {
        return (
          <li className={styles.listItem} key={id}>
            <div>{type}</div>
            <div className={styles.value}>{value}</div>
            <div>{time}</div>
          </li>
        );
      })}
    </ul>
  );
}
