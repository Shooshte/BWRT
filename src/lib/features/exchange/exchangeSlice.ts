import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

import { mockTrades } from "./trades";

export interface Trade {
  amount: number;
  id: string;
  price: number;
  timestamp: number;
  type: "buy" | "sell";
}

interface ChartState {
  BTC: number;
  USD: number;
  trades: Trade[];
}

const initialState: ChartState = {
  BTC: 0.11342278,
  USD: 225.05,
  trades: mockTrades,
};

export const chartSlice = createSlice({
  name: "exchange",
  initialState,
  reducers: {
    addTrade: (state, action: PayloadAction<Trade>) => {
      state.trades = [action.payload, ...state.trades];
    },
    setBTC: (state, action: PayloadAction<number>) => {
      state.BTC = action.payload;
    },
    setUSD: (state, action: PayloadAction<number>) => {
      state.USD = action.payload;
    },
  },
});

export const { addTrade, setBTC, setUSD } = chartSlice.actions;
export default chartSlice.reducer;
