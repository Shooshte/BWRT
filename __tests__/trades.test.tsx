import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

import { mockTrades } from "../src/lib/features/exchange/trades";
import Trades from "../src/app/components/trades";
import { round } from "../src/lib/utils";
import { Wrapper } from "../src/tests/test-utils";

describe("Trades", () => {
  test("Render correct DOM structure", () => {
    const { container } = render(<Trades />, { wrapper: Wrapper });
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("Render correct trades", () => {
    mockTrades.forEach(({ amount, id, price, timestamp, type }) => {
      it(`render trade ${id}`, () => {
        render(<Trades />, { wrapper: Wrapper });
        const trade = screen.getByTestId(`trade-${id}`);
        expect(trade).toBeInTheDocument();

        const roundedAmount = round(amount, 5);
        const roundedPrice = round(price * amount, 2);

        const value =
          type === "buy"
            ? `${roundedAmount} BTC / - ${roundedPrice} USD`
            : `- ${roundedAmount} BTC / + ${roundedPrice} USD`;
        const time = new Date(timestamp).toLocaleTimeString();

        expect(trade).toHaveTextContent(type);
        expect(trade).toHaveTextContent(value);
        expect(trade).toHaveTextContent(time);
      });
    });
  });
});
