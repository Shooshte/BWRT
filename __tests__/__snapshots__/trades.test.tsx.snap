// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trades Render correct DOM structure 1`] = `
<ul
  class="list"
>
  <li
    class="listItem"
    data-testid="trade-4"
  >
    <div>
      buy
    </div>
    <div
      class="value"
    >
      0.00018 BTC / - 191.46 USD
    </div>
    <div>
      6:03:02 PM
    </div>
  </li>
  <li
    class="listItem"
    data-testid="trade-7"
  >
    <div>
      sell
    </div>
    <div
      class="value"
    >
      - 0.002 BTC / + 2121.34 USD
    </div>
    <div>
      6:03:01 PM
    </div>
  </li>
  <li
    class="listItem"
    data-testid="trade-9"
  >
    <div>
      buy
    </div>
    <div
      class="value"
    >
      0.001 BTC / - 1063.71 USD
    </div>
    <div>
      6:02:42 PM
    </div>
  </li>
  <li
    class="listItem"
    data-testid="trade-3"
  >
    <div>
      buy
    </div>
    <div
      class="value"
    >
      0.00017 BTC / - 180.88 USD
    </div>
    <div>
      1:36:35 PM
    </div>
  </li>
  <li
    class="listItem"
    data-testid="trade-2"
  >
    <div>
      sell
    </div>
    <div
      class="value"
    >
      - 0.00012 BTC / + 129.07 USD
    </div>
    <div>
      1:36:24 PM
    </div>
  </li>
  <li
    class="listItem"
    data-testid="trade-1"
  >
    <div>
      buy
    </div>
    <div
      class="value"
    >
      0.00015 BTC / - 161.27 USD
    </div>
    <div>
      1:36:22 PM
    </div>
  </li>
  <li
    class="listItem"
    data-testid="trade-5"
  >
    <div>
      sell
    </div>
    <div
      class="value"
    >
      - 0.0001 BTC / + 107.5 USD
    </div>
    <div>
      1:35:24 PM
    </div>
  </li>
</ul>
`;
